ghw/libghw.c

```diff
@@ -367,11 +420,23 @@ ghw_read_str (struct ghw_handler *h)
 
   if (hdr[0] != 0 || hdr[1] != 0 || hdr[2] != 0 || hdr[3] != 0)
     return -1;
-  h->nbr_str = ghw_get_i32 (h, &hdr[4]);
+
+  // Read number of strings and total string size. Make sure that nbr_string
+  // is not too large, as it's incremented afterwards.
+  h->nbr_str = (uint32_t) ghw_get_i32 (h, &hdr[4]);
+  if (h->nbr_str >= UINT32_MAX-8) ghw_error_exit();
   h->nbr_str++;
-  h->str_size = ghw_get_i32 (h, &hdr[8]);
-  h->str_table = (char **) malloc ((h->nbr_str + 1) * sizeof (char *));
-  h->str_content = (char *) malloc (h->str_size + h->nbr_str + 1);
+  h->str_size = (uint32_t) ghw_get_i32 (h, &hdr[8]);
+
+  // Allocate str_table, str_content. Avoid exceeding size_t limits.
+  uint64_t alloc_size = ((uint64_t)h->nbr_str) * sizeof (char *);
+  if (alloc_size > SIZE_MAX) ghw_error_exit();
+  h->str_table = (char **) malloc_unwrap (alloc_size);
```

Comment:
I think it's clearer/safer to use `calloc`.

Reply:
Agreed, fixed.